text/0006-monolothic-packaging.md

```diff
@@ -0,0 +1,389 @@
+
+## Poor Ergonomics
+
+In addition to the peer npm dependency issue described above, due to the fact
+that CDK is "hyper modular", in all languages, we have poor ergonomics when it
+comes to declaring dependencies. Since users are required to explicitly install
+a module for each service they use, even simple projects end up with dozens of
+direct CDK dependencies.
+
+# Basic Example
+
+The AWS CDK will be shipped as a single module that includes the core types and the entire AWS Construct Library.
+
+This means that a 3rd-party library will declare its dependency on the CDK via a single module `aws-cdk-lib` (name to be decided):
+
+```json
+{
+  "name": "your-awesome-lib",
+  "peerDependencies": {
+    "aws-cdk-lib": "^2.12.0",
+    "constructs": "^1.0.0"
+  }
+}
+```
+
+> The `constructs` module includes the core programming model of the CDK, which we plan to extract into a separate library.
+
+And that's it. Apps that consume this library will depend on it and will also
+depend on the CDK:
+
+```json
+{
+  "name": "my-awesome-app",
+  "dependencies": {
+    "aws-cdk-lib": "2.89.0",
+    "constructs": "1.77.0",
+    "your-awesome-lib": "^1.0.0"
+  }
+}
+```
+
+In JavaScript code, import statements that use the CDK will now look like this:
+
+```ts
+import { aws_s3, aws_dynamodb } from 'aws-cdk-lib';
+import { Construct } from 'constructs';
+
+// ...
+export class MyConstruct extends Construct {
+  constructor(scope: Construct, id: string) {
+    super(scope, id);
+
+    new aws_s3.Bucket(this, ...);
+    new aws_dynamodb.Table(this, ...);
+  }
+}
+```
+
+> TODO: other languages
+
+# Design Summary
+
+The general approach proposed by this RFC is to ship the core types and the
+entire AWS Construct Library as a single module for all languages.
+
+This will dramatically simplify how users declare their compatibility with the
+AWS CDK, and I would argue it is also more aligned with our user's mental model (see [Rational](#rationale-and-alternatives) beflow).
+
+# Detailed Design
+
+See [monocdk-experiment](https://www.npmjs.com/package/monocdk-experiment) for a protoype.
+
+## Submodules
+
+The monolithic module will be organized into submodules that match 1:1 the current module system we have for the AWS CDK. These submodules will be implemented using typescript **namespaced exports** (see [jsii PR](https://github.com/aws/jsii/pull/1297)).
+
+* All `@aws-cdk/core` types will be exported without a namespace (root).
+* Hyphens in the current module names will be converted to underscores (`aws-s3` => `aws_s3`). 
+
+### Should we use the `aws_` prefix?
+
+We still need some way to discern between AWS Construct Library modules such as `aws-s3` and framework modules such as `assets`. If we didn't have the `aws-` prefix, and an AWS service will be introduced with the name of a module we already have (e.g. say we launched an "AWS Assets" service one day), we would have a conflict. 
+
+## Issues with Specific Modules
```

Comment:
Is the intention that the MonoCDK combines everything currently under the `@aws-cdk/aws-` scope in the aws-cdk repo currently, or is it just for L1s and L2s?

If it's the latter, might want to add discussion on edge packages like `@aws-cdk/aws-lambda-ndoe`. And if it's the former, might want to clarify that it includes those packages.

Reply:
added